### DISCUSSION TITLE
Error when prerendering page that gets data from api

### DISCUSSION THREAD

**Discussion by ZeroThe2nd** [vercel/next.js#44723] [Help] [answered]
Using the new `app` directory, I'm writing a page that does a fetch call to one of my API endpoints made in `/pages/api`.

I read in #43415 you'll need to provide a full, non-relative URL. This works just fine locally by adding a `API_URL` environment variable that contains `http://127.0.0.1:3000`, making the API calls to my local dev server. However when attempting to deploy to Vercel, it all goes down south.

Vercel errors out on `Unexpected token < in JSON` errors, likely because it somehow receives an HTML page instead of the JSON output from the API. This while I have configured my `API_URL` variable dynamically in the vercel config like so:

```js
env: {
  API_URL: process.env.VERCEL_URL ? `https://${process.env.VERCEL_URL}` : process.env.PUBLIC_URL
},
```
Whenever it's deploying on vercel, the `VERCEL_URL` variable exists, and will be used. According to the `console.log` I added as a test, it is using the dynamic URL for vercel. However it seems as if the page prerender happens _before_ the API route(s) are generated, thus giving them some other response.

Would anyone happen to have a suggestion to fix this problem, thus making API routes work properly?

**Comment by harblaith7** [comment]
I always get this. I followed the docs and I still get this error. It might be a bug

**Comment by icyJoseph** [comment, marked as answer]
This is not an issue exclusive to the app dir, the same would happen if you were trying to reach for an api route from `getStaticProps` - Since there's no server during build time, there's no way to handle API routes.

In general reaching out for pages/api is a bit of a catch-22, specially when pages are pre-rendered. Do remember that Next.js, both in pages and appDir, tries to pre-render by default.

So what can we do? Well the simplest way is to parametrize whatever work you do inside pages/api, into a re-usable function, and call that directly from your server component, or GetStaticProps/GetServerSideProps function.

Why?

In order for you to have `pages/api` on VERCEL_URL, you ought to succeed deploying once, once you succeed, the next time you build, you'll reach out for this, now, old version of `pages/api` pre-render with that, effectively using it as a 3rd party input, and then deploy again. This is the catch-22 you are falling into. 

There's a reason why the absolute URL requirement is there, but it is nothing that can't be solved by properly making reusable functions that can be called from within Server Components, GetStaticProps, or whatever other server side context.

Don't believe me? Take it from the docs: https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

It applies just as well to a Server Components model, or `getStaticProps`.

**Reply by ZeroThe2nd** [reply]
Handling the API stuff differently did resolve the issue, but indeed that doesn't resolve the "using your own API Routes" thing. Which is a framework thing on its own.

Thanks for the enlightenment however! I'm able to continue the work now. m(._.)m

**Reply by icyJoseph** [reply]
I do recall that in Nuxt, there is a helper function that can take a string like `"/api/something"` and at build time, it mocks a request, IIRC, it doesn't matter how, but it calls the api function for you. https://nuxt.com/docs/guide/concepts/server-engine#direct-api-calls
